### PR TITLE
apkguard bugfix

### PR DESCRIPTION
**Pull request by enovella**
Found the bug (https://github.com/rednaga/APKiD/blame/master/apkid/rules/dex/packers.yara#L88) with a new sample.

Using @radare, we can see the Dalvik asm: (highlighted the byte that should be ??)
apkguard-fix

Previous rule:
`1a0a 2f00       // const-string v10, "lllllllllllllllllllllllllllllllllllllllll.zip"`

New rule:
`1a0a 3000       // const-string v10, "lllllllllllllllllllllllllllllllllllllllll.zip"`

Final rule
`1a0a ??00       // const-string v10, "lllllllllllllllllllllllllllllllllllllllll.zip"`

### CODE REVIEW

**Comment by CalebFenton** [comment]
Does this represent a new version? If so, it'd be nice to capture that with a new rule. This may be a hard question to answer. If you can't be sure, let's prefer this:

Is it the case that our current rule was too strict and wasn't taking into account all possible permutations?

**Comment by enovella** [comment]
As the name represents, it is a bufix on Apkguard. I am afraid that @strazzere was a bit strict when creating the rule. No new versions.

**Comment by strazzere** [comment]
This one is interesting. I think I did include this by accident so it's likely a bug. Though it'd be interesting to figure out what the true difference is between the stubs. Since, if this rule doesn't hit, it means the string table shifted a bit due to more strings (though most of the ones I've looked at had identical string tables) - though this would cause a miss if they ever added one or removed one string (or just swapped some I guess?).

TLDR - bug, but maybe it makes it pinned to a specific version of apkguard.

**Comment by enovella** [comment]
As discussed over Slack, the sample I obfuscated was a little bit different (40 dex files or so). Although, only the classes.dex contains the unpacker code, so it should be due to a string table with more offsets to new strings.

Have you checked these samples? https://github.com/rednaga/APKiD/pull/68#issuecomment-351803111

**Comment by enovella** [comment]
Here you go:
```
0x00002634      1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2325 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x000025e8      1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22e9 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x000025e0      1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22e3 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x000025fc      1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2303 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x00002554      1a0a3d00       const-string v10, str.zzmsztZMFtywFBuhygAMTHFRwVtIz.zip ; 0x2418 ; "!zzmsztZMFtywFBuhygAMTHFRwVtIz.zip"
0x000025f4      1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22e1 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x000025bc      1a0a0100       const-string v10, str.BznoThaWwXIiJEtTXputrtkJDKwGSqF.zip ; 0x39c ; "#BznoThaWwXIiJEtTXputrtkJDKwGSqF.zip"
0x000025fc      1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22ed ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x00002664      1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2350 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x00002624      1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2311 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x00002580      1a0a2000       const-string v10, str.VhseDgIpUAWyMjgaariQjxdhNjcdrtZDazK.zip ; 0x225b ; "'VhseDgIpUAWyMjgaariQjxdhNjcdrtZDazK.zip"
0x00002624      1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2318 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x00002548      1a0a3700       const-string v10, str.kcGpHeMJvpwrGhPukIHiJJVYGymwRu.zip ; 0x23af ; "\"kcGpHeMJvpwrGhPukIHiJJVYGymwRu.zip"
0x00002640      1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2341 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x0000263c      1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x233c ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
0x00002598      1a0a1f00       const-string v10, str.WSmFCuZqrzrHiNLNolsSyEKHLJWrXVZ.zip ; 0x2254 ; "#WSmFCuZqrzrHiNLNolsSyEKHLJWrXVZ.zip"
```

**Comment by enovella** [comment]
Opcodes are ordered now:
```c
1a0a0100       const-string v10, str.BznoThaWwXIiJEtTXputrtkJDKwGSqF.zip ; 0x39c ; "#BznoThaWwXIiJEtTXputrtkJDKwGSqF.zip"
1a0a1f00       const-string v10, str.WSmFCuZqrzrHiNLNolsSyEKHLJWrXVZ.zip ; 0x2254 ; "#WSmFCuZqrzrHiNLNolsSyEKHLJWrXVZ.zip"
1a0a2000       const-string v10, str.VhseDgIpUAWyMjgaariQjxdhNjcdrtZDazK.zip ; 0x225b ; "'VhseDgIpUAWyMjgaariQjxdhNjcdrtZDazK.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22e1 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22ed ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2311 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2318 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2325 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2350 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22e3 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22e9 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2303 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x233c ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2341 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3700       const-string v10, str.kcGpHeMJvpwrGhPukIHiJJVYGymwRu.zip ; 0x23af ; "\"kcGpHeMJvpwrGhPukIHiJJVYGymwRu.zip"
1a0a3d00       const-string v10, str.zzmsztZMFtywFBuhygAMTHFRwVtIz.zip ; 0x2418 ; "!zzmsztZMFtywFBuhygAMTHFRwVtIz.zip"1a0a0100       const-string v10, str.BznoThaWwXIiJEtTXputrtkJDKwGSqF.zip ; 0x39c ; "#BznoThaWwXIiJEtTXputrtkJDKwGSqF.zip"
1a0a1f00       const-string v10, str.WSmFCuZqrzrHiNLNolsSyEKHLJWrXVZ.zip ; 0x2254 ; "#WSmFCuZqrzrHiNLNolsSyEKHLJWrXVZ.zip"
1a0a2000       const-string v10, str.VhseDgIpUAWyMjgaariQjxdhNjcdrtZDazK.zip ; 0x225b ; "'VhseDgIpUAWyMjgaariQjxdhNjcdrtZDazK.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22e1 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22ed ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2311 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2318 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2325 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a2f00       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2350 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22e3 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x22e9 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2303 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x233c ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3000       const-string v10, str.lllllllllllllllllllllllllllllllllllllllll.zip ; 0x2341 ; "-lllllllllllllllllllllllllllllllllllllllll.zip"
1a0a3700       const-string v10, str.kcGpHeMJvpwrGhPukIHiJJVYGymwRu.zip ; 0x23af ; "\"kcGpHeMJvpwrGhPukIHiJJVYGymwRu.zip"
1a0a3d00       const-string v10, str.zzmsztZMFtywFBuhygAMTHFRwVtIz.zip ; 0x2418 ; "!zzmsztZMFtywFBuhygAMTHFRwVtIz.zip"
```

**Comment by enovella** [comment]
Hi Guys,

do you think that APKGuard (v.1) always uses the zip filenames `lllllllllllllllllllllllllllllllllllllllll.zip`? Thus, opcodes would be always either `1a0a2f00`  or `1a0a3000`.

Perhaps, the new CryptoShell (v.2) uses random zip filenames and thus the opcodes can vary a bit. 

Would you like to split that into two rules?

**Comment by CalebFenton** [comment]
How about splitting it into three rules? The first is rule is what we have already, the second rule is the `1a0a 3000`, and the third is a general rule with the string index generalized to `??` (what you refer to as the final rule).

It would be best to map these different rules to specific versions, but that's not always possible and it's rarely easy. So in lieu of that, we can make a note in the comments that we're not sure exactly what the version is, but the name itself could be something like "APKGuard (legacy)" for the first rule, "APKGuard (CryptoShell)" for the second rule, and "APKGuard (generic)" for the third rule. The third, generic rule, should only match if the first two don't match.

**Comment by enovella** [comment]
To be honest, I personally dislike having 3 heavy rules for just detecting a tiny modification. I believe that this is not very efficient., isn't it?

**Comment by CalebFenton** [comment]
I don't think the rules would be heavy. Most of the other rule "families" follow a similar development. There's one main rule with most of the logic, and then several smaller, lighter rules which specifically identify a version. The goal with apkid is to provide as much "intelligence" about the protection as possible. If we can tell the difference between older and newer versions of a tool, that may be useful. Small details accrete into larger inferences, and it's probably worth it to capture anything we've discovered with the rules.

If we ever run into performance problems, we can consolidate rules or offer a minimized rule set.

**Comment by enovella** [comment]
Sure @CalebFenton! 

APKGuard is only `(2f|30)`, otherwise it will be CryptoShell. Yeah, it makes a lot of sense.

**Comment by strazzere** [comment]
This LGTM, thanks for figuring that one out and patching the rule 👍
